### PR TITLE
feat(android): continue playback with media controls and touch zoom

### LINKED ISSUE #1085: feat(android): add locked-screen media controls and touch player polish
Follow-up to #1069 and the Android beta lifecycle work merged in #1070.

Implementation: #1091

## Problem

Android playback stops or loses useful controls when the screen locks. The app also lacks an Android media notification tied to the currently presented video. Switching away from a playing video tab can leave Picture-in-Picture black, and Android does not expose the existing setting that keeps the video visible as a cross-tab mini-player. In mobile fullscreen, the bottom-right action dock can cover Shaka submenus and SponsorBlock notices even though desktop already coordinates the action dock with menus, end screens, notices, and fullscreen side docks. The existing video zoom control has mouse and keyboard interactions but no touch-native pinch gesture. Switching a playing video to the audio-only format for true background playback forces Shaka to replace its media source, which can cause an audible 0.4 to 1 second interruption.

### Locked-screen playback and media notification

- [x] Add an Android-only setting that allows playback to continue when the screen locks or the app backgrounds. Enable it by default.
- [x] When the setting is disabled, keep the current pause/background behavior.
- [x] Publish the presented tab's title, channel, artwork, duration, position, and playback state through an Android media session.
- [x] Show an Android media notification with working play, pause, seek, and stop controls. Add previous and next controls only when the active queue or playlist can perform those actions.
- [x] Route lock-screen, headset, Bluetooth, and notification actions to the presented player without allowing a hidden background tab to take ownership.
- [x] Keep playback alive with the Android mechanism required for active media playback. Do not claim survival after process death or force-stop.
- [x] Release native media-session and foreground-service state promptly when playback stops, ends, or the owning tab closes.
- [x] Preserve Picture-in-Picture, audio focus, notification permission, and multi-tab behavior.
- [x] Measure the background transition and accept the current 0.4–1-second audio-only handoff for #1091; track a seamless handoff separately in #1094.

### Cross-tab mini-player and Picture-in-Picture

- [x] Expose the existing "Keep a mini player on screen when switching tabs" setting on Android.
- [x] Keep the playing video visible as a detached mini-player when the user switches tabs and the setting is enabled.
- [x] Make "Return to video tab" select the owning Capacitor tab.
- [x] Keep Android media controls owned by the visible detached mini-player without letting an unrelated background tab take control.
- [x] Enter Picture-in-Picture with the detached player's live video instead of a black frame.

### Mobile fullscreen layering

- [x] Hide or move the bottom-right fullscreen action dock out of the way whenever a Shaka submenu is open, following the existing submenu-open and submenu-close behavior used by custom controls.
- [x] Keep SponsorBlock notices readable and actionable instead of drawing them behind the action dock.
- [x] Preserve the existing desktop layering for SponsorBlock notices, end screens, Shaka menus, and fullscreen side docks.
- [x] Cover phone and tablet layouts, both display directions, and non-100% UI scales.

### Touch video zoom

- [x] Add two-finger pinch-to-zoom to the existing video zoom feature.
- [x] Clamp and persist gesture changes through the existing zoom model and supported range.
- [x] Keep the gesture centered on the pinch focal point and preserve panning of zoomed video.
- [x] Do not let a pinch trigger tap-to-toggle controls, double-tap seek, fullscreen swipe, menu actions, or page scrolling.

### Terminal SponsorBlock prompts and Chapters state

- [x] Dismiss prompt-to-skip SponsorBlock notices when playback reaches the end instead of leaving their media-time countdown frozen.
- [x] Highlight the Chapters player-option tile while the chapters panel is open.

## Verification

- [x] On the Pixel 8 Pro running Android 16, video continued through screen lock and remained controllable through the media notification and lock-screen play/pause controls; wake-lock and media-session ownership were also verified.
- [x] Verify the physical audio-only background path on the Pixel during screen lock.
- [x] Repeat physical-device media-button coverage with a wired headset and Bluetooth device.
- [x] Verify at the automated renderer/native boundary that disabling locked-screen playback pauses all players when Android backgrounds the app and clears native media-session state.
- [x] Verify media-notification ownership across presented and detached tab switches, Picture-in-Picture, and task removal on the Pixel and at the automated renderer/native boundary.
- [x] Complete physical-device notification coverage for unload, tab close, queue changes, and normal playback completion.
- [x] Accept for #1091 that the audio-only background handoff can buffer for roughly 0.4–1 second; reliable screen-lock playback is preferred for now, and seamless transitions are tracked in #1094.
- [x] Enable the cross-tab mini-player, switch away from a playing video, return through its action, and press Home from the detached player. The detached video and PiP should remain live instead of turning black.
- [x] Verify in phone/tablet E2E coverage that open Shaka submenus and SponsorBlock notices remain above the mobile action dock and tappable.
- [x] Pinch in and out over the video at several focal points. Zoom and pan should remain bounded, responsive, and isolated from one-finger player gestures.
- [x] Add regression tests at the renderer-to-native media boundary and the actual multi-tab, PiP, fullscreen, and touch behavior boundaries.

- [x] On API 35, verify the assembled APK removes an active SponsorBlock prompt on the player ended event and gives the expanded Chapters tile the active option color.

## Physical Android follow-up, 2026-09-04

Tested the current development APK on a Pixel 8 Pro running Android 16:

- A Soundcore Liberty 5 Bluetooth gesture paused and resumed playback while the headset was the active A2DP route.
- A Bowers & Wilkins Px7 S2e USB headset button paused and resumed playback while the headset was the active wired route.
- Adding a recommended video to the watch queue changed the Android media action bitmap from 335 to 367, adding Next. Removing the item restored 335.
- Unloading the playing tab and closing the reloaded playing tab both removed the media session and notification.
- Normal playback completion at 4:55 showed Replay, then removed the media session and notification.

### Current coverage

- Pixel 8 Pro on Android 16: video playback through screen lock, media notification play/pause, wake lock and media-session ownership, pinch gestures, exact zoom percentage persistence, service/notification teardown after removing the task, visible cross-tab mini-player, working return action, and live Picture-in-Picture from the detached player. A direct no-format-switch experiment kept the same blob URL and `readyState` 4, but Android paused the video about three seconds after display-off, so true locked-screen playback still requires an audio-only handoff or another native-safe equivalent. The resulting short buffer is accepted for #1091 and tracked separately in #1094.
- API 35 emulator: renderer playback, media session, notification actions, and steady-playback notification update behavior.
- API 24 emulator: native foreground media service and four-action `MediaStyle` notification. Its factory WebView 69 cannot parse the current renderer bundle. After installing Chrome 119 and selecting it as the WebView provider, the same API 24 emulator loaded the current renderer normally. On supported physical devices, users can likewise update Android System WebView or Chrome through a trusted app store; OpenTubeX cannot bundle or replace the system WebView itself.
- Automated coverage: renderer/native ownership and opt-out behavior, multi-tab and detached-mini-player isolation, Picture-in-Picture target isolation, touch focal anchoring and tap isolation, fractional UI scale, and mobile fullscreen menu/SponsorBlock/action-dock layering.

The verification matrix above is complete for the tested Pixel and emulators. Android versions and OEMs beyond those devices remain additional coverage. Playback remains alive only while the renderer process and foreground media service exist. It does not resume after process death, reboot, task removal, or force-stop.

### PR DESCRIPTION
**Pull request by D3SOX**
## Pull Request Type

- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Related issue

Closes #1085

## Description

Android playback currently loses useful system controls when the app is hidden or the screen locks. Switching away from a playing video tab can also leave Picture-in-Picture black, and Android hides the existing cross-tab mini-player preference. Mobile fullscreen controls can overlap Shaka menus and SponsorBlock notices, and video zoom has no touch gesture.

This adds a foreground Android media service and session for the presented player, with lock-screen and notification controls, artwork, position, conditional queue navigation, and a wake lock while playing. A device-local Player setting, enabled by default, lets users opt out and pause playback whenever the app is hidden. Playing video switches to audio-only while hidden and restores its video stream when shown again.

Android now exposes "Keep a mini player on screen when switching tabs." A visible detached mini-player retains media-session ownership, its return action selects the owning Capacitor tab, and entering Picture-in-Picture from it keeps the live video visible. Hidden background tabs still cannot take over controls.

The player now supports bounded, focal-point-aware pinch zoom and panning. It claims Android touch input while zoom gestures are enabled so Chromium cannot cancel a human pinch before the second finger lands. Fullscreen Shaka submenus hide the mobile action dock, while SponsorBlock notices remain above the dock and below open menus.

SponsorBlock prompt-to-skip notices now close when playback ends instead of freezing their media-time countdown on the final second. The Chapters tile also uses its expanded state to show the same active color as other selected mobile player options.

## Release note category

- [x] Not noteworthy
- [ ] Highlights
- [ ] More improvements
- [ ] Fixed bugs

## Release note

Android can now continue playback with lock-screen media controls and a cross-tab mini-player, while the mobile player adds pinch-to-zoom and keeps fullscreen menus and SponsorBlock notices clear of the action dock.

## Testing

1. On Android, play a video, press Home, and lock the screen. Audio should continue, and the Android media notification and lock screen should control only the presented player.
2. In Player Settings, disable "Continue playback when the screen is locked". Hiding the app should now pause every playing tab and remove the media notification.
3. Enable "Keep a mini player on screen when switching tabs", play a video, and switch to another tab. The video should remain visible. Its return action should select the video tab, and pressing Home while it is detached should show live video in Picture-in-Picture instead of a black frame.
4. With a queue or playlist, verify that previous and next appear only when those actions are available. Closing the owning tab or swiping the app out of Recents should remove the notification and stop background playback.
5. Pinch in and out over the video, including over the paused center control and from an already panned zoom level. The focal point should stay anchored, and the next tap should still toggle playback.
6. In mobile fullscreen, open each Shaka submenu and trigger a SponsorBlock notice. Menus and notices should remain visible and tappable without the action dock covering them.

7. On a phone layout, open Chapters from the player options and verify its tile stays highlighted while the panel is open. Let a prompt-to-skip SponsorBlock segment reach the video end; the prompt should close instead of remaining on its final second.

## Additional context

The durability boundary is an active renderer plus its Android foreground media service. This keeps playback alive after backgrounding and screen lock, but it does not reconstruct or resume playback after process death, reboot, or force-stop. Removing the task intentionally stops the service because no renderer remains to own the media session. The audio-only background handoff can buffer for roughly 0.4–1 second; that is accepted for this PR in favor of reliable locked-screen playback, with seamless transitions tracked in #1094.

Coverage includes a Pixel 8 Pro on Android 16 and API 35 and API 24 emulators. On the Pixel, the detached mini-player stays visible, its return action works, its media notification retains the correct owner, and Picture-in-Picture shows live video. The API 24 native service and four-action MediaStyle notification work. Its factory WebView 69 cannot parse the current renderer bundle, while installing Chrome 119 and selecting it as the WebView provider made the same emulator load the renderer normally. On supported physical devices, Android System WebView or Chrome can be updated through a trusted app store; OpenTubeX cannot bundle or replace the system WebView itself. Other OEM and Android-version behavior has not been claimed from the Pixel result.

The assembled APK was also checked in the API 35 WebView: an expanded Chapters tile changed from the neutral background to the active color, and an end-of-video SponsorBlock prompt was removed on the player ended event. The Pixel was offline for this final follow-up build.

### CODE REVIEW

**Comment by D3SOX** [comment]
@coderabbitai rate limit

**Comment by D3SOX** [comment]
@coderabbitai review

File: android/app/src/main/java/org/opentubex/app/AndroidMediaSessionService.java [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+                artwork = loaded;
+                applyState(currentState);
+            });
+        });
+    }
+
+    private Bitmap downloadArtwork(String source) {
+        HttpURLConnection connection = null;
+        try {
+            connection = (HttpURLConnection) new URL(source).openConnection();
```
[the remark this answers is not in the document]
Comment (D3SOX) [reply android/app/src/main/java/org/opentubex/app/AndroidMediaSessionService.java]: > [!NOTE]  \n> gpt-5.4 responding on behalf of Nico\n\nFixed in 8987782f7. Artwork downloads now require HTTPS without URL credentials, reject loopback, link-local, private, carrier-grade NAT, multicast, unspecified, and IPv6 unique-local destinations, disable automatic redirects, and revalidate every redirect target. Android JVM coverage exercises public and blocked address classes.

File: src/renderer/App.css [hunk 2]
```diff
@@ -163,6 +163,12 @@
   touch-action: pan-x;
 }
 
+/* Chromium may claim the first finger as a pan before the second finger lands,
+   cancelling the pointer stream before the player can recognize a pinch. */
+.app.capacitorTabs :deep(.ftVideoPlayer.videoZoomTouchEnabled:not(.scrollMiniPlayer)) {
+  touch-action: none;
```
[the remark this answers is not in the document]
Comment (D3SOX) [reply src/renderer/App.css:169]: > [!NOTE]  \n> gpt-5.4 responding on behalf of Nico\n\nI am keeping this selector as written. The Pixel reproduced Chromium claiming the first touch before a second finger landed, which cancelled the custom pinch.  is required before the gesture begins; adding it only after the second pointer or limiting it to fullscreen makes inline pinch unreliable again. The player already used  for the mobile fullscreen-swipe gesture, so this does not newly remove vertical page scrolling from inline video. The setting and issue explicitly require touch pinch on the video, including inline playback.

File: src/renderer/components/ft-shaka-video-player/ft-shaka-video-player.js [hunk 3]
```diff
@@ -5551,21 +5713,24 @@ export default defineComponent({
...
 
+    function registerMediaSessionHandlers() {
       tabMediaCoordinator.setActionHandlers(mediaTabId, 'player', {
         play: () => video.value?.play(),
         pause: () => video.value?.pause(),
         stop: () => {
           const videoElement = video.value
           if (!videoElement) return
+          const wasPaused = videoElement.paused
+          mediaSessionStopped = true
```
[the remark this answers is not in the document]
Comment (D3SOX) [reply src/renderer/components/ft-shaka-video-player/ft-shaka-video-player.js:5730]: > [!NOTE]  \n> gpt-5.4 responding on behalf of Nico\n\nFixed in 8987782f7. A video ID change now clears , and  clears it only after the inactive-tab guard accepts playback.

File: src/renderer/tabs/TabMediaCoordinator.js [hunk 4]
```diff
@@ -36,6 +35,10 @@ function getEntry(tabId) {
 }
 
 function chooseOwner() {
+  if (process.env.IS_CAPACITOR) {
+    return presentedTabId && mediaByTabId.has(presentedTabId) ? presentedTabId : null
+  }
+
```
[the remark this answers is not in the document]
Comment (D3SOX) [reply src/renderer/tabs/TabMediaCoordinator.js:45]: > [!NOTE]  \n> gpt-5.4 responding on behalf of Nico\n\nFixed in 8987782f7 with the cross-tab mini-player case preserved. Android immediately drops outgoing action ownership, then after Vue settles it pauses the outgoing player before clearing the native session when no mini-player is visible. If that player became the visible cross-tab mini-player, it remains the media-session owner instead. Unit coverage locks down both branches.

File: src/renderer/App.css (hunk 2 above)
[the remark this answers is not in the document]
Comment (D3SOX) [reply src/renderer/App.css:169]: > [!NOTE]  
> gpt-5.4 responding on behalf of Nico

Correction to my previous reply: the required declaration is `touch-action: none`; the earlier mobile player rule was `touch-action: pan-x`. The Pixel reproduced Chromium claiming the first touch before the second finger landed, so fullscreen-only gating would make inline pinch unreliable.

File: src/renderer/components/ft-shaka-video-player/ft-shaka-video-player.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (D3SOX) [reply src/renderer/components/ft-shaka-video-player/ft-shaka-video-player.js:5730]: > [!NOTE]  
> gpt-5.4 responding on behalf of Nico

Correction to my previous reply: a video ID change now clears `mediaSessionStopped`, and `handlePlay` clears it only after the inactive-tab guard accepts playback. Fixed in 8987782f7.

**Comment by D3SOX** [comment]
@coderabbitai rate limit

**Comment by D3SOX** [comment]
@coderabbitai rate limit

**Comment by D3SOX** [comment]
@coderabbitai rate limit

**Comment by D3SOX** [comment]
@coderabbitai review

File: src/renderer/components/ft-shaka-video-player/ft-shaka-video-player.js [hunk 5]
```diff
@@ -10595,6 +10800,11 @@ export default defineComponent({
 
       cleanUpCustomPlayerControls()
 
+      videoZoomTouchPointers.clear()
+      videoZoomPinchStart = null
+      videoZoomGestureZoom.value = null
+      clearTimeout(videoZoomSuppressClickTimer)
+
```
[the remark this answers is not in the document]
Comment (D3SOX) [reply src/renderer/components/ft-shaka-video-player/ft-shaka-video-player.js:10807]: > [!NOTE]  
> gpt-5.4 responding on behalf of Nico

Fixed in 78ec70a17. Player teardown now releases its mini-player registration before removing its media handlers and playback state. The coordinator regression covers that teardown sequence and verifies stale actions are no longer routed to the removed mini-player.

File: android/app/src/main/java/org/opentubex/app/AndroidMediaSessionService.java [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+                    int status = connection.getResponseCode();
+                    if (isRedirectStatus(status)) {
+                        String location = connection.getHeaderField("Location");
+                        if (location == null || redirects == MAX_ARTWORK_REDIRECTS) return null;
+                        url = new URL(url, location);
+                        continue;
+                    }
+                    if (status < 200 || status >= 300) return null;
+                    try (InputStream input = connection.getInputStream()) {
+                        return BitmapFactory.decodeStream(input);
```
[the remark this answers is not in the document]
Comment (D3SOX) [reply android/app/src/main/java/org/opentubex/app/AndroidMediaSessionService.java]: > [!NOTE]  
> gpt-5.4 responding on behalf of Nico

Fixed in 78ec70a17. Artwork responses are now capped at 5 MiB even when Content-Length is absent or inaccurate. Decoding first probes dimensions, calculates a bounded power-of-two sample size, requests ARGB_8888 without density scaling, and verifies the resulting 2048px / 4-Mpx / 16-MiB limits. JVM regressions cover declared and streamed byte overflows plus oversized-image sampling.
